compiler/rustc_ty_utils/src/abi.rs

```diff
@@ -480,13 +480,22 @@ fn fn_abi_sanity_check<'tcx>(
                         // The unstable abi `PtxKernel` also uses Direct for now.
                         // It needs to switch to something else before stabilization can happen.
                         // (See issue: https://github.com/rust-lang/rust/issues/117271)
-                        assert!(
-                            matches!(&*tcx.sess.target.arch, "wasm32" | "wasm64")
-                                || matches!(spec_abi, ExternAbi::PtxKernel | ExternAbi::Unadjusted),
-                            "`PassMode::Direct` for aggregates only allowed for \"unadjusted\" and \"ptx-kernel\" functions and on wasm\n\
-                          Problematic type: {:#?}",
-                            arg.layout,
-                        );
+
+                        match spec_abi {
+                            // The unadjusted ABI is ill specified, but unfortunately we need it for
+                            // calling certain LLVM intrinsics.
+                            ExternAbi::Unadjusted => {}
+                            ExternAbi::PtxKernel => {}
+                            ExternAbi::C { unwind: _ }
+                                if matches!(&*tcx.sess.target.arch, "wasm32" | "wasm64") => {}
+                            _ => {
+                                panic!(
+                                    "`PassMode::Direct` for aggregates only allowed for \"unadjusted\" and \"ptx-kernel\" functions and on wasm\n\
+                                      Problematic type: {:#?}",
+                                    arg.layout,
+                                );
+                            }
+                        }
```

Comment:
No, this function depends on having access to TyCtxt. In particular `tcx.struct_tail_for_codegen()`. As for these specific exclusions, I don't think we should. This is a ratchet to prevent regressions in usage of `PassMode::Direct`. It will be gone in the not too far future hopefully.

Reply:
That's fair then.